plugins/api/api_main.ml

```diff
@@ -1,39 +1,215 @@
+  let of_string x = Path.reduce @@ Path.filename_of_string x
+  let to_string = Path.string_of_filename
+  let basename = Path.basename
+  let basename' x = to_string @@ basename @@ of_string x
+
+  let exists p = FileUtil.(test Exists) @@ to_string p
+  let is_apiable = FileUtil.(Or (Is_file, Is_link))
+
+  let is_apiable_path p =
+    exists p && FileUtil.test is_apiable @@ to_string p
+
+  let find_files path =
+    try_with_default ~default:[]
+      (fun () -> FileUtil.find is_apiable (to_string path)
+          (fun acc f -> of_string f :: acc) [])
+
+  let path_of_desc desc = of_string desc.lang / of_string desc.name
+  let with_lang path lang = path / of_string lang
+
+  let lang_of_path path =
+    to_string Path.(path |> reduce |> dirname |> basename)
+
+  let name_of_path p = to_string @@ Path.basename p
+  let desc_of_path p = {name = name_of_path p; lang = lang_of_path p;}
+
+  let api_of_path path =
+    if is_apiable_path path then
+      Some {desc = desc_of_path path; path;}
+    else None
+
+  let is_api_at_path desc path = is_apiable_path @@ path / path_of_desc desc
+  let find_all_paths paths desc = List.filter ~f:(is_api_at_path desc) paths
+  let find_path paths desc = List.find ~f:(is_api_at_path desc) paths
+
+  let find_apis ?lang path =
+    let path' = match lang with
+      | Some lang -> with_lang path lang
+      | None -> path in
+    find_files path' |>
+    List.filter_map ~f:api_of_path
+
+end
+
+module Api = struct
+
+  type t = api
+  type op_err = [ `No_dst | `Fail of exn ]
+
+  let lang t = t.desc.lang
+  let name t = t.desc.name
+
+  let create lang path =
+    let path = Api_path.of_string path in
+    if Api_path.is_apiable_path path then
+      let name = Api_path.name_of_path path in
+      Some {desc = {lang; name}; path;}
+    else None
+
+  let cp dest api =
+    let dest' = Api_path.with_lang dest (lang api) in
+    ok_if_true (Api_path.exists dest) ~error:`No_dst >>= fun () ->
+    try_with
+      (fun () ->
+         let dest' = Api_path.to_string dest' in
+         FileUtil.mkdir dest';
+         FileUtil.cp [Api_path.to_string api.path] dest')
+
+  let rm path desc =
+    let path = Api_path.(path / path_of_desc desc) in
+    ok_if_true (Api_path.exists path) ~error:`No_dst >>= fun () ->
+    try_with (fun () -> FileUtil.rm [Api_path.to_string path])
+
+end
+
+module Api_error = struct
+
+  let exit_with () = exit 1
+
+  let api_exists path desc =
+    exit_with @@ eprintf
+      "Will not overwrite an API named %s at %s. Remove it before the updating.\n"
+      desc.name (Api_path.to_string path)
+
+  let api_not_exists ?path desc = match path with
+    | Some path ->
+      exit_with @@ eprintf "Can't find api %s:%s at %s\n"
+        desc.lang desc.name (Api_path.to_string path)
+    | None ->
+      exit_with @@ eprintf "Can't find api %s:%s\n"
+        desc.lang desc.name
+
+  let no_path api =
+    exit_with @@ eprintf
+      "There are not any writable paths for %s plugin\n" api.name
```

Comment:
Put it simpler - error messages are for users, not for programmers:

`"Can't store an api %s: permission denied" `